pages/v4/Coroutines.md

```diff
@@ -0,0 +1,184 @@
+---
+title: "Coroutines"
+category: "basics"
+displayOrder: 10
+version: "v4"
+---
+
+First of all, what is a Coroutine?
+
+Taken from Wikipedia:
+
+_Coroutines are very similar to threads. However, coroutines are cooperatively multitasked, whereas threads are typically preemptively multitasked. This means that coroutines provide concurrency but not parallelism. The advantages of coroutines over threads are that they may be used in a hard-realtime context (switching between coroutines need not involve any system calls or any blocking calls whatsoever), there is no need for synchronisation primitives such as mutexes, semaphores, etc. in order to guard critical sections, and there is no need for support from the operating system._
+
+A Coroutine is a piece of code that executes "in parallel", over a certain amount of time, or generally asynchronously, with another; in our particular case, it is code that should run for an arbitrarily long amount of frames, without getting in the way of the usual flow of your code.
+
+This **does not mean** that the Coroutine is operating on a separate thread, and thus can be written ignoring the usual constraints of the Update-Draw loop of the game engine; each step of the Coroutine should be able to be performed in one Update call, so don't go around looking for prime numbers in a Coroutine, chances are your game would stutter.
+
+# A simple example
+
+Let's say, for example, that you want to code a Component that, upon pressure of a key, will make your GameObject's move 200 units to the right: with only this requirement you will probably end up with something like this
+
+```csharp
+public void OnUpdate()
+{
+	if (DualityApp.Keyboard.KeyHit(Key.Space))
+	{
+		this.GameObj.Transform.Pos.MoveBy(new Vector2(200, 0));
+	}
+}
+```
+
+Which is fine, and works well for you.
+But now, let's assume you want the transition to be smooth; one way would be to move 1 unit per frame (let's ignore the time it would take or the fact that it would be dependent on the frame rate)
+
+```csharp
+private Vector2? movementTarget = null;
+private Vector2 movementDirection = Vector2.Zero;
+
+public void OnUpdate()
+{
+	if(movementTarget != null)
+	{
+		if((this.movementTarget.Value - this.GameObj.Transform.Pos.Xy).LengthSquared >= this.movementDirection.LengthSquared)
```

Comment:
Code style: Missing whitespace after the `if` and before the `(`, same in the line above.